### PR TITLE
[buster] + Apply example_ynh

### LINKED ISSUE #56: [buster] Install fails because dependency libjemalloc1 ain't available anymore, should be replaced by libjemalloc2 I believe
This dependency was added back to https://github.com/YunoHost-Apps/discourse_ynh/commit/5be413e2dd754c1b45531ecb89b5e1b037420cd5

Unfortunately this package ain't available for buster (c.f. [here](https://packages.debian.org/stretch/libjemalloc1), lack of 'buster' link in top right corner) but there's a similar one on buster which is called [libjemalloc2](https://packages.debian.org/buster/libjemalloc2) (but I have no idea what/why this dependency is needed by discourse so can't really say for sure, should look at the install instructions I guess)

Anyway, if this is the right solution, possibly just using the 'or' syntax to require either of package should be the appropriate fix I suppose.

### PR DESCRIPTION
**Pull request by yalh76**
and adding libjemalloc2 for buster for Fix #56

## PR Status
- [x] Code finished.
- [x] Tested with Package_check.
- [x] Fix or enhancement tested.
- [x] Upgrade from last version tested.
- [x] Can be reviewed and tested.

## Package_check results
---
[Build Status/badge/icon)](https://ci-apps-dev.yunohost.org/jenkins/job/discourse_ynh%20PR57%20(yalh76)/)  
Failed cause of: `The underlying system error is Errno::EDQUOT: Disk quota exceeded @ io_write -`

### Stretch => [Logs](https://paste.yunohost.org/bakegobide.bash)
image

### Buster => [Logs](https://paste.yunohost.org/dukanejemi.sql)
image

### CODE REVIEW

**Comment by maniackcrudelis** [comment]
@yalh76 do you know if we really need to match the exact number of workers in the log at this [line](https://github.com/YunoHost-Apps/discourse_ynh/blob/fix-buster/scripts/backup#L87)

Asking that because `unicorn_workers` is defined in the install script but not stored.
So the backup script fails at that point.

And also because maybe there's a line for each worker that have started and so the idea was to wait for the last one.
My CI is currently running (hard) so I can't have a look to that log right now.

**Comment by JimboJoe** [comment]
I hadn't found another way to check if all workers were started...

**Comment by maniackcrudelis** [comment]
An install from a not particularly fast VM gave me that:
```
I, [2020-04-28T15:44:50.024797 #11918]  INFO -- : worker=0 ready
I, [2020-04-28T15:44:50.987545 #11926]  INFO -- : worker=1 ready
I, [2020-04-28T15:44:51.350307 #11934]  INFO -- : worker=2 ready
```
1s and a half to get from the first one to the last one.
Yet 1s and a half is enough for a 502...

The backup script is not tested right away, the remove script goes before.
Yet I think the best would probably to do that properly and store the value defined during the install.

**Comment by maniackcrudelis** [comment]
Sorry @JimboJoe, saw your comment to late.
Just read the log, and I don't think there would be any other way. I was thinking about a regex for the number. But as said in my previous comment, I think it's better to get the last one the way you did.

**Comment by maniackcrudelis** [comment]
So, I'm going to implement that on this branch, so the variable $unicorn_workers will goes with the package and be reused everywhere it's needed.

**Comment by maniackcrudelis** [comment]
4 hours after I've started to test the app, I'm still waiting for Package check to finish and tell me if everything is ok with my last commit...

**Comment by maniackcrudelis** [comment]
And finally...
```
Info:  Restoring the systemd configuration...
Warning: yunohost.hook <lambda> - [3998.1] Failed to enable unit: Unit file discourse.service does not exist.
```

**Comment by maniackcrudelis** [comment]
Thanks @yalh76 for the clear example of why `ynh_restore` should never be used...
I read the diff of this PR yesterday without noticing it, reading the stable restore script right now, I spot the error.
https://github.com/YunoHost-Apps/discourse_ynh/blob/master/scripts/restore#L54

And actually the backup script have a backup of the systemd script, https://github.com/YunoHost-Apps/discourse_ynh/blob/master/scripts/backup#L57

Reading that part, https://github.com/YunoHost-Apps/discourse_ynh/blob/master/scripts/restore#L87-L92, I was wondering how Discourse was managing its systemd script !

**Comment by maniackcrudelis** [comment]
Ok, now @yalh76 it works for me.
Although I didn't tested on stretch, but I do not expect it to fail.

**Comment by yalh76** [comment]
@maniackcrudelis thanks for the work done. I'm also waiting for package_check results.
It was just a first draft to apply example_ynh started a week ago

**Comment by JimboJoe** [comment]
Thanks a lot for that huge work! :heart: 
I took time to review (minor typo fixes), and made extensive tests, including upgrading from previous version and ARM installation. I fixed edge cases for change_url and restore, and updated README to remove obsolete known limitations.

**Comment by yalh76** [comment]
Sorry it was a first try, made with a blackhoe.... 
So thanks you for the review and the commits and thanks to @maniackcrudelis for the tests and commits.

For now, package check are level7 for stretch and buster.

The only point I left behind is in the upgrade script: there was some conditions I removed:
https://github.com/YunoHost-Apps/discourse_ynh/blob/086de718f49a506c66498d2f7610f5fd0ce07d50/scripts/upgrade#L83

I don't know if you seen them. if yes, let's merge ;) If not, I can work on them.

PS: I packaged several apps and it's the first time I wasn't able to understand how to package an app when reading the docker_composer files, it's jut a mess.... So +10 for you @JimboJoe, you are like a yunohost packaging yoda

File: scripts/upgrade [hunk 1]
```diff
@@ -357,7 +357,7 @@ ynh_script_progression --message="Securing files and directories..."
 
 # Add a pids and socket directory for the systemd script.
 mkdir -p "$final_path/tmp/pids"
-mkdir "$final_path/tmp/sockets"
+mkdir -p "$final_path/tmp/sockets"
```
Comment (JimboJoe) [inline scripts/upgrade:360]: `$final_path/tmp` is created the line before, so how come that changes anything...? :thinking:

Comment (yalh76) [reply scripts/upgrade:360]: because when upgrading from  2.3.10~ynh1, `$final_path/tmp/sockets` is already present so it fails to create it and the upgrade fails :'(

**Comment by yalh76** [comment]
Do we merge ?

**Comment by alexAubin** [comment]
Yolomerging because we're closing in to buster release and this PR is asleep though apparently working ?

**Comment by alexAubin** [comment]
(Can still tweak stuff during testing I guess)
